# What kind of data does pandas handle?

I want to start using pandas

In [37]:
import pandas as pd

To load the pandas package and start working with it, import the package. 

The community agreed alias for pandas is pd, so loading pandas as pd is assumed standard practice for all of the pandas documentation.

## pandas data table representation

In [38]:
df = pd.DataFrame(
    {
        "Name": [
            "Braund, Mr. Owen Harris",
            "Allen, Mr. William Henry",
            "Bonnell, Miss. Elizabeth",
        ],
        "Age": [22, 35, 58],
        "Sex": ["male", "male", "female"],
    }
)


A DataFrame is a 2-dimensional data structure that can store data of different types (including characters, integers, floating point values, categorical data and more) in columns. 

It is similar to a spreadsheet, a SQL table or the data.frame in R.

The table has 3 columns, each of them with a column label. The column labels are respectively Name, Age and Sex.

The column Name consists of textual data with each value a string, the column Age are numbers and the column Sex is textual data.

In spreadsheet software, the table representation of our data would look very similar:

In [39]:
df

Name  Age     Sex
0   Braund, Mr. Owen Harris   22    male
1  Allen, Mr. William Henry   35    male
2  Bonnell, Miss. Elizabeth   58  female

To manually store data in a table, create a DataFrame. 

When using a Python dictionary of lists, the dictionary keys will be used as column headers and the values in each list as columns of the DataFrame.

In [40]:
df["Age"]

0    22
1    35
2    58
Name: Age, dtype: int64

Each column in a DataFrame is a Series

you can create a Series from scratch as well:

In [41]:
ages=pd.Series([22,23,24,25], name="Age")

In [42]:
ages

0    22
1    23
2    24
3    25
Name: Age, dtype: int64

A pandas Series has no column labels, as it is just a single column of a DataFrame. A Series does have row labels.

# Do something with a DataFrame or Series

I want to know the maximum Age of the passengers

We can do this on the DataFrame by selecting the Age column and applying max():



In [43]:
df["Age"].max()

58

In [44]:
ages.max()

25

As illustrated by the max() method, you can do things with a DataFrame or Series. 

pandas provides a lot of functionalities, each of them a method you can apply to a DataFrame or Series. 

As methods are functions, do not forget to use parentheses ().

In [45]:
df.describe()

Age
count   3.000000
mean   38.333333
std    18.230012
min    22.000000
25%    28.500000
50%    35.000000
75%    46.500000
max    58.000000

The describe() method provides a quick overview of the numerical data in a DataFrame. 

As the Name and Sex columns are textual data, these are by default not taken into account by the describe() method.

Many pandas operations return a DataFrame or a Series. The describe() method is an example of a pandas operation returning a pandas Series or a pandas DataFrame.

# How do I read and write tabular data?

In [46]:
df=pd.read_csv("G:\data_science\data\iris.csv")

pandas provides the read_csv() function to read data stored as a csv file into a pandas DataFrame. pandas supports many different file formats or data sources out of the box (csv, excel, sql, json, parquet, …), each of them with the prefix read_*.

In [47]:
df.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

use the head() method with the required number of rows (in this case default 5 rows are shown) as argument. 

To see the first N rows of a DataFrame


In [48]:
df.tail()

sepal.length  sepal.width  petal.length  petal.width    variety
145           6.7          3.0           5.2          2.3  Virginica
146           6.3          2.5           5.0          1.9  Virginica
147           6.5          3.0           5.2          2.0  Virginica
148           6.2          3.4           5.4          2.3  Virginica
149           5.9          3.0           5.1          1.8  Virginica

use the Tails() method with the required number of rows (in this case default 5 rows are shown) as argument. 

To see the Last N rows of a DataFrame


In [49]:
df.dtypes

sepal.length    float64
sepal.width     float64
petal.length    float64
petal.width     float64
variety          object
dtype: object

When asking for the dtypes, no brackets are used! dtypes is an attribute of a DataFrame and Series. 

Attributes of DataFrame or Series do not need brackets. 

Attributes represent a characteristic of a DataFrame/Series, whereas a method (which requires brackets) do something with the DataFrame/Series

In [50]:
df.to_excel("iris.xlsx", sheet_name="data", index=False)

Whereas read_* functions are used to read data to pandas, the to_* methods are used to store data. 

The to_excel() method stores the data as an excel file. In the example here, the sheet_name is named passengers instead of the default Sheet1. 

By setting index=False the row index labels are not saved in the spreadsheet.

The equivalent read function read_excel() will reload the data to a DataFrame:

In [51]:
df=pd.read_excel("iris.xlsx", sheet_name="data")

In [52]:
df.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal.length  150 non-null    float64
 1   sepal.width   150 non-null    float64
 2   petal.length  150 non-null    float64
 3   petal.width   150 non-null    float64
 4   variety       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


The method info() provides technical information about a DataFrame, so let’s explain the output in more detail:

It is indeed a DataFrame.

There are 150 entries, i.e. 150 rows.

Each row has a row label (aka the index) with values ranging from 0 to 149.

The table has 5 columns. Most columns have a value for each of the rows (all 150 values are non-null).

The columns sepal.length, Sepal.width, petal.length and petal.width consists of numerical data (strings, aka object). The other columns like variety consiat of textual data.

The kind of data (characters, float,…) in the different columns are summarized by listing the dtypes.

The approximate amount of RAM used to hold the DataFrame is provided as well.

# How do I select a subset of a DataFrame?

### How do I select specific columns from a DataFrame?

In [54]:
sepal_width=df["petal.width"]

In [55]:
sepal_width.head()

0    0.2
1    0.2
2    0.2
3    0.2
4    0.2
Name: petal.width, dtype: float64

To select a single column, use square brackets [] with the column name of the column of interest

Each column in a DataFrame is a Series. 

As a single column is selected, the returned object is a pandas Series.

 We can verify this by checking the type of the output:

In [57]:
type(df["sepal.width"])

pandas.core.series.Series

In [58]:
df["sepal.width"].shape

(150,)

DataFrame.shape is an attribute of a pandas Series and DataFrame containing the number of rows and columns: (nrows, ncolumns). A pandas Series is 1-dimensional and only the number of rows is returned

# How do I filter specific rows from a DataFrame?

In [62]:
Setosa=df[df["variety"]=='Setosa']

To select rows based on a conditional expression, use a condition inside the selection brackets [].

In [64]:
Setosa.head()

sepal.length  sepal.width  petal.length  petal.width variety
0           5.1          3.5           1.4          0.2  Setosa
1           4.9          3.0           1.4          0.2  Setosa
2           4.7          3.2           1.3          0.2  Setosa
3           4.6          3.1           1.5          0.2  Setosa
4           5.0          3.6           1.4          0.2  Setosa

The condition inside the selection brackets df["variety"]=='Setosa' checks for which rows the variety column has a value equal to Setosa.

The output of the conditional expression (>, but also ==, !=, <, <=,… would work) is actually a pandas Series of boolean values (either True or False) with the same number of rows as the original DataFrame.

Such a Series of boolean values can be used to filter the DataFrame by putting it in between the selection brackets []. 

Only rows for which the value is True will be selected.

In [65]:
Setosa.shape

(50, 5)

# How do I select specific rows and columns from a DataFrame?

### 1. Using **.loc**:

.loc is primarily label-based indexing, which means you can select rows and columns by their labels (row and column names).

In [71]:
titanic=pd.read_csv(r"G:\data_science\data\titanic.csv")

In [72]:
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [73]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


I’m interested in the names of the passengers older than 35 years.

In [74]:
adult_names=titanic.loc[titanic["Age"]>35,"Name"]

In [76]:
adult_names.head()

1     Cumings, Mrs. John Bradley (Florence Briggs Th...
6                               McCarthy, Mr. Timothy J
11                             Bonnell, Miss. Elizabeth
13                          Andersson, Mr. Anders Johan
15                     Hewlett, Mrs. (Mary D Kingcome) 
Name: Name, dtype: object

In this case, a subset of both rows and columns is made in one go and just using selection brackets [] is not sufficient anymore. The loc/iloc operators are required in front of the selection brackets []. When using loc/iloc, the part before the comma is the rows you want, and the part after the comma is the columns you want to select.

### 2. Using .iloc:
.iloc is primarily integer-location based indexing, which means you can select rows and columns by their integer positions.
python

import pandas as pd

#### Assume you have a DataFrame named 'df'

#### Select a single row by index
selected_row = df.iloc[row_index]

#### Select multiple rows by indexes
selected_rows = df.iloc[row_index_list]

#### Select rows and specific columns by indexes
selected_rows_cols = df.iloc[row_index, column_index_list]

In [77]:
titanic.iloc[2:50,3:8]

Name     Sex   Age  SibSp  \
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   
20                               Fynney, Mr. Joseph J    male  35.0      0   
21                              Beesley, Mr. Lawrence    male  34.0      0   
22                        McGowan, Miss. Anna "Annie"  female  15.0      0   
23                       Sloper, Mr. William Thompson    male  28.0      0   
24                      Palsson, Miss. Torborg Danira  female   8.0      3   
25  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
26                            Emir, Mr. Farred Chehab    male   NaN      0   
27                     Fortune, Mr. Charles Alexander    male  19.0      3   
28                      O'Dwyer, Miss. Ellen "Nellie"  female   NaN      0   
29                                Todoroff, Mr. Lalio    male   NaN      0   
30                           Uruchurtu, Don. Manuel E    male  40.0      0   
31     Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
32                           Glynn, Miss. Mary Agatha  female   NaN      0   
33                              Wheadon, Mr. Edward H    male  66.0      0   
34                            Meyer, Mr. Edgar Joseph    male  28.0      1   
35                     Holverson, Mr. Alexander Oskar    male  42.0      1   
36                                   Mamee, Mr. Hanna    male   NaN      0   
37                           Cann, Mr. Ernest Charles    male  21.0      0   
38                 Vander Planke, Miss. Augusta Maria  female  18.0      2   
39                        Nicola-Yarred, Miss. Jamila  female  14.0      1   
40     Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.0      1   
41  Turpin, Mrs. William John Robert (Dorothy Ann ...  female  27.0      1   
42                                Kraeff, Mr. Theodor    male   NaN      0   
43           Laroche, Miss. Simonne Marie Anne Andree  female   3.0      1   
44                      Devaney, Miss. Margaret Delia  female  19.0      0   
45                           Rogers, Mr. William John    male   NaN      0   
46                                  Lennon, Mr. Denis    male   NaN      1   
47                          O'Driscoll, Miss. Bridget  female   NaN      0   
48                                Samaan, Mr. Youssef    male   NaN      2   
49      Arnold-Franchi, Mrs. Josef (Josefine Franchi)  female  18.0      1   

    Parch  
2       0  
3       0  
4       0  
5       0  
6       0  
7       1  
8       2  
9       0  
10      1  
11      0  
12      0  
13      5  
14      0  
15      0  
16      1  
17      0  
18      0  
19      0 

When specifically interested in certain rows and/or columns based on their position in the table, use the iloc operator in front of the selection brackets [].

In [79]:
air_quality=pd.read_csv(r"https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/air_quality_no2.csv", index_col=0,parse_dates=True)

In [80]:
air_quality.to_csv("airquality.csv")